### PR TITLE
Automatic differentiation

### LINKED ISSUE #66: Support abs, min, max in expressions + automatic differentiation
## Details:
Hours Spent: 0

## Background
The plotter should support also functions such as `abs`, `min` and `max`.  These operators lead to functions that are not differentiable everywhere but differentiable almost everywhere.  It is not feasible to represent the derivative of these functions in our expression language.

## Overview
- Introduce operators abs, min ans max into the expression language.
- Replace differentiation of expression with an evaluator that returns a value of an expression together with the (interval) derivative and second derivative (or its rough approximation or interval enclosure).

[Wikipedia](https://en.wikipedia.org/wiki/Automatic_differentiation)

## Acceptance Criteria
- [x] The plotter accepts input such as "abs(x)" or "max(x,-x)".
- [x] The plotter plots both rough graphs and enclosures of such functions.
- [x] The quality of smooth plots is not decreased.

## Closing Notes

Overview of changes can be found in PR #113

### PR DESCRIPTION
**Pull request by michalkonecny**
# Issues
Closes #66 

# Summary
- Replace symbolic differentiation with automatic differentiation
- Support abs, min, max in expressions
- Improve segmentation algorithm to work also for abs, min, max in typical cases
- Add Approx min, max
- Fix Approx sine for very large exact arguments
- Fix Approx log for non-exact arguments

### CODE REVIEW

File: src/Plot/RobustPlot.purs [hunk 1]
```diff
@@ -54,35 +75,58 @@ plotEnclosures canvasSize bounds domainSegments evaluator = segmentEnclosures
...
+      xMidPoint = centreA x
+
+      enclosureWidth = xUA - xLA
+
+      twoA = fromRationalPrec 50 two
+
+      evaluatorX = evaluator2 x
 
-      xGradGrad = map boundsA $ evaluator.f'' x
+      xMidPointValue = boundsA <$> (evaluator xMidPoint <#> (_.value))
```
Comment (RelativeForce) [inline src/Plot/RobustPlot.purs]: Should this be written as `boundsA <$> (_.value)  <$>  evaluator xMidPoint` to be consistent with the rest of the file

File: src/Plot/RobustPlot.purs [hunk 2]
```diff
@@ -54,35 +75,58 @@ plotEnclosures canvasSize bounds domainSegments evaluator = segmentEnclosures
...
+
+      xGradGrad = boundsA <$> (evaluatorX <#> (_.derivative2))
```
Comment (RelativeForce) [inline src/Plot/RobustPlot.purs]: Same as the above comment

File: src/Plot/RobustPlot.purs [hunk 3]
```diff
@@ -137,42 +181,28 @@ plotEnclosures canvasSize bounds domainSegments evaluator = segmentEnclosures
...
-    canvasXLower = toCanvasX xLower
-
-    canvasXUpper = toCanvasX xUpper
 
   toCanvasX :: Number -> Number
   toCanvasX x = ((x - xLowerBound) * canvasWidth) / rangeX
 
-  toCanvasY :: Approx -> Number
-  toCanvasY yApprox = safeCanvasY
+  -- toCanvasY :: Approx -> Number
```
Comment (RelativeForce) [inline src/Plot/RobustPlot.purs]: Remove zombie code

**Review by RelativeForce — APPROVED** [review]
Looks fine just some minor changes

File: src/Plot/RobustPlot.purs (hunk 2 above)
Comment (michalkonecny) [reply src/Plot/RobustPlot.purs]: Good point, thank you.

File: src/Plot/RobustPlot.purs (hunk 3 above)
Comment (michalkonecny) [reply src/Plot/RobustPlot.purs]: Done

File: src/Plot/RobustPlot.purs (hunk 1 above)
Comment (michalkonecny) [reply src/Plot/RobustPlot.purs]: Done, thank you.

**Pull request merged**